src/lib/proxy-server.ts

```diff
@@ -0,0 +1,190 @@
+import { platform } from 'os';
+import { promisify } from 'util';
+import httpProxy from 'http-proxy';
+import sudo from 'sudo-prompt';
+import { loadUserData } from 'src/storage/user-data';
+
+const sudoExec = promisify( sudo.exec );
+
+let proxyServer: http.Server | null = null;
+let isProxyRunning = false;
+
+/**
+ * Gets the port number for a given domain by looking it up in user data
+ */
+async function getPortForDomain( domain: string ): Promise< number | null > {
+	try {
+		const userData = await loadUserData();
+		// Find the site with the matching custom domain
+		const site = userData.sites.find(
+			( site ) => site.useCustomDomain && site.customDomain === domain
+		);
+		return site ? site.port : null;
+	} catch ( error ) {
+		console.error( 'Error looking up domain in user data:', error );
+		return null;
+	}
+}
+
+/**
+ * Attempts to start the proxy server on port 80
+ * This requires admin/root privileges
+ */
+export async function startProxyServer(): Promise< boolean > {
+	if ( isProxyRunning ) return true;
+
+	try {
+		// First, try starting on a privileged port directly (might work if app has permissions)
+		const result = await startDomainProxy();
+		if ( result ) {
+			return true;
+		}
+
+		// If direct start fails and we're not skipping elevation, attempt with sudo/admin privileges
+		return await startProxyWithElevation();
+	} catch ( error ) {
+		console.error( 'Failed to start proxy server:', error );
+		return false;
+	}
+}
+
+/**
+ * Attempts to start the proxy server directly (without privilege elevation)
+ * @param {number} port - The port to listen on (defaults to 80)
+ */
+async function startDomainProxy( port: number = 80 ): Promise< boolean > {
+	try {
+		// Create proxy with additional options to preserve host header
+		const proxy = httpProxy.createProxyServer();
+
+		proxy.on( 'error', ( err, req, res ) => {
+			console.error( 'Proxy error:', err );
+			if ( res && 'writeHead' in res ) {
+				res.writeHead( 500 );
+				res.end( 'Proxy error: ' + err.message );
+			}
+		} );
+
+		proxyServer = http.createServer( async ( req, res ) => {
+			const host = req.headers.host?.split( ':' )[ 0 ]; // Remove port if present
+
+			// Log incoming request information for debugging
+			console.log( `Received request with host header: ${ host }` );
+
+			// Look up the port directly from user data
+			if ( ! host ) {
+				console.log( 'No host header found' );
+				res.writeHead( 404 );
+				res.end( 'No host header found' );
+				return;
+			}
+
+			const port = await getPortForDomain( host );
+			if ( ! port ) {
+				console.log( `Domain not found: ${ host }` );
+				res.writeHead( 404 );
+				res.end( `Domain not found: ${ host }` );
+				return;
+			}
+
+			console.log( `Proxying request for ${ host } to port ${ port }` );
+
+			// Forward the request with the original host preserved
+			proxy.web( req, res, {
+				target: `http://localhost:${ port }`,
+				xfwd: true, // Pass along x-forwarded headers
+			} );
+		} );
```

Comment:
```suggestion
		proxyServer = http.createServer( async ( req, res ) => {
			const host = req.headers.host?.split( ':' )[ 0 ]; // Remove port if present

			// Look up the port directly from user data
			if ( ! host ) {
				console.log( 'No host header found' );
				res.writeHead( 404 );
				res.end( 'No host header found' );
				return;
			}

			const port = await getPortForDomain( host );
			if ( ! port ) {
				console.log( `Domain not found: ${ host }` );
				res.writeHead( 404 );
				res.end( `Domain not found: ${ host }` );
				return;
			}

			// Forward the request with the original host preserved
			proxy.web( req, res, {
				target: `http://localhost:${ port }`,
				xfwd: true, // Pass along x-forwarded headers
			} );
		} );
```
Let's remove the console.log's here